helix-event/src/runtime.rs

```diff
@@ -0,0 +1,88 @@
+//! The event system makes use of global to decouple different systems.
+//! However, this can cause problems for the integration test system because
+//! it runs multiple helix applications in parallel. Making the globals
+//! thread-local does not work because a applications can/does have multiple
+//! runtime threads. Instead this crate implements a similar notion to a thread
+//! local but instead of being local to a single thread, the statics are local to
+//! a single tokio-runtime. The implementation requires locking so its not exactly efficient.
```

Comment:
So if I'm understanding this right, during integration tests, there's only one process running the whole show, and we're building and invoking `Application`s directly in the tests, each of which runs in parallel in different threads. And because of the combination of these things, combined with having a single global static registry, we get weird conflicts with event registry because multiple threads are trying to initialize the same set of events and hooks into the same map?

Reply:
yeah pretty much. Initializing the same set of events multiple times is not really as much of a problem (that happens anyway as some integration tests start multiple application instances after one another), the problem would be concurrently running integration tests messing with each other.

For example if all integration tests used the same registry and test `A` fired a `DocumentDidChange` event hooks from test `B would be triggered. Similarly the job queue is now a static aswell. Jobs are used for lots of things (like writing files as you know well :D) and if different integration tests werent' isolated that would mean that they would share a job queue (so one test may steal a job intended for another test).

 The code in this module prevents that. An alternative could be to run integration tests on a single thread and make this a thread local during integration tests but that would prevent testing multitheraded behaviour (and also would involve every integration test) so I opted for this approach instead

Reply:
Ohh damn, I didn't even consider the global job pool interaction. I kind of wonder if it would be simpler just to fork off a child process for integration tests instead of running in the same process.

Reply:
dealing with subprocesses is also kind of pain (they don't really work well with the rust test system). If we want to avoid the complexity here, changing all integration tests to use a single-threaded runtime is probably the simplest solution (but would require a change to every integration test)

Reply:
Yeah, I'm seeing that after some research. This is unfortunate. I think the reason every test uses the multithreaded runtime is because it panics if we don't, though I can't remember why. And even if that weren't a problem, it would still end up being much slower. This seems like an ok workaround unless we can think of something else.